pipeline/astro_cal.py

```diff
@@ -28,11 +149,472 @@ class AstroCalibrator:
+                        '-CROSSID_RADIUS', str( crossid_rad )
+                       ]
+
+            # TODO : use a different gaia magnitude for different image
+            # filters (see Issue #107)
+
+            t0 = time.perf_counter()
+            if catexp.origin == 'GaiaDR3':
+                command.extend( [ '-ASTREFMAG_KEY', 'MAG_G', '-ASTREFMAGERR_KEY', 'MAGERR_G' ] )
+            else:
+                raise NotImplementedError( f"Don't know what magnitude key to choose for astrometric reference "
+                                           f"{catexp.origin}; only GaiaDR3 is implemented." )
+
+            res = subprocess.run( command, capture_output=True )
+            t1 = time.perf_counter()
+            _logger.debug( f"Scamp with {sources.num_sources} sources and {catexp.num_items} catalog stars "
+                           f"(with match_nmax={max_nmatch}) took {t1-t0:.2f} seconds" )
+            if res.returncode != 0:
+                raise SubprocessFailure( res )
+
+            scampstat = astropy.io.votable.parse( xmlfile ).get_table_by_index( 1 )
+            nmatch = scampstat.array["AstromNDets_Reference"][0]
+            sig0 = scampstat.array["AstromSigma_Reference"][0][0]
+            sig1 = scampstat.array["AstromSigma_Reference"][0][1]
+            infostr = ( f"Scamp on {pathlib.Path(catfile).name} to catalog with nominal magnitude "
+                        f"range {catexp.minmag}-{catexp.maxmag} yielded {nmatch} matches out of "
+                        f"{len(sources.data)} sources and {len(catexp.data)} catalog objects, "
+                        f"with position sigmas of ({sig0:.2f}\", {sig1:.2f}\")" )
+            if not ( ( nmatch > self.pars.min_frac_matched * min( len(sources.data), len(catexp.data ) ) )
+                     and ( nmatch > self.pars.min_matched_stars )
+                     and ( ( sig0 + sig1 ) / 2. <= self.pars.max_arcsec_residual )
+                    ):
+                infostr += ( f", which isn't good enough.\n"
+                             f"Scamp command: {res.args}\n"
+                             f"-------------\nScamp stderr:\n{res.stderr.decode('utf-8')}\n"
+                             f"-------------\nScamp stdout:\n{res.stdout.decode('utf-8')}\n" )
+                # A warning not an error in case something outside is iterating
+                _logger.warning( infostr )
+                raise BadMatchException( infostr )
+
+            _logger.info( infostr )
+
+            # Move the header information written in the ".head" file
+            # scamp created to the image header, and to a WCS object
+            # we're going to return.
+            with open( headfile ) as ifp:
+                hdrtext = ifp.read()
+            # The FITS spec says ASCII, but Emmanuel put a non-ASCII latin-1
+            # character in his comments... and astropy.io.fits.Header is
+            # anal about only ASCII.  Sigh.
+            hdrtext = hdrtext.replace( 'é', 'e' )
```

Comment:
It'd be awfully nice if we could move to a modern file format like HDF5, but there is SO MUCH infrastructure that uses FITS that we really just have to accept a lot of this legacy stuff.  (Not just things like sextractor, scamp, etc., but also astropy's file handling is pretty tightly tied to FITS.)

Reply:
I know, it is a real tragedy. When we started the W-FAST codebase we were lucky that we needed fast I/O so we did everything in HDF5 (and then snuck in some native MATLAB files because it was so easy) but I think that this is the right way to go in the future but the momentum is very big to stay with FITS (like a lot of things in astronomy).